### ISSUE TITLE
Goss bagging dependent on nthread

### ISSUE THREAD

**Issue by CHY-WANG** [microsoft/LightGBM#5441]
## Description
Samples being selected by bagging when using 'boosting_type': 'goss' can vary depending on the value of 'nthread', even if all other settings are exactly the same.
 
A reproducible example is included in the end.

## Environment info

LightGBM version: v3.3.2
Python: 3.8.5

## Additional Comments

The bagging is done by blocks of sampels.

In https://github.com/microsoft/LightGBM/blob/v3.3.2/include/LightGBM/utils/threading.h#L44
[include/LightGBM/utils/threading.h](https://github.com/microsoft/LightGBM/blob/v3.3.2/include/LightGBM/utils/threading.h#L44),
`    *out_nblock = std::min<int>(
        num_threads,
        static_cast<int>((cnt + min_cnt_per_block - 1) / min_cnt_per_block));
    if (*out_nblock > 1) {
      *block_size = (cnt + (*out_nblock) - 1) / (*out_nblock);
      // force the block size to the times of min_cnt_per_block
      *block_size = (*block_size + min_cnt_per_block - 1) / min_cnt_per_block *
                    min_cnt_per_block;
    } else {
      *block_size = cnt;
    }`
The number of threads can affect the number of blocks, and thus the sizes of the blocks. Depending on num_thread, block_size can be n*min_cnt_per_block where n can be any positive integer depending on num_thread.

Then when goss is doing bagging within a block in [src/boosting/goss.hpp](https://github.com/microsoft/LightGBM/blob/master/src/boosting/goss.hpp#L103),
all the n*min_cnt_per_block samples are considered when determining the bagging result. For example the threshold for selecting samples with top gradients is determined by the gradients of all n*min_cnt_per_block samples, which depdends on num_thread.

An easy fix I can think of will be just set
` *out_nblock = static_cast<int>((cnt + min_cnt_per_block - 1) / min_cnt_per_block);`
then nothing will depend on num_thread any more. Or maybe there is other better solution.

## Reproducible example

The example is in Python.
Code:
`import lightgbm
import numpy as np
import pandas as pd

# Generate some toy data
train_data = pd.DataFrame({"y":np.zeros(1024*8)})
predictors = []
for i in range(1000):
    train_data["x" + str(i)] = np.sin(np.array(range(1024*8))+i/1000) + 1
    train_data["y"] = train_data["y"] = train_data["x" + str(i)]
    predictors.append("x" + str(i))
valid_data = train_data.copy()
train_data_lgb = lightgbm.Dataset(train_data[["x1", "x2"]],
                                  label = train_data["y"],
                                  free_raw_data = False)
valid_data_lgb = lightgbm.Dataset(valid_data[["x1", "x2"]],
                                  label = valid_data["y"],
                                  free_raw_data = False)

# Use goss
params = {'boosting_type': 'goss',
          'cat_l2': 55.08312413819303,
          'feature_fraction': 0.7245736321334011,
          'learning_rate': 0.09327701408526637,
          'max_cat_threshold': 26,
          'max_depth': 4,
          'min_data_in_leaf': 35,
          'min_gain_to_split': 0.37881681266148276,
          'other_rate': 0.14278147196648264,
          'reg_alpha': 4.1101593522818165,
          'reg_lambda': 65.70629589637446,
          'top_rate': 0.30250663837184166,
          'bagging_fraction': 1.0,
          'bagging_freq': 0,
          'bagging_seed': 121,
          'feature_fraction_seed': 214,
          'feature_pre_filter': False,
          'num_iterations': 100,
          'num_leaves': 127,
          'seed': 214,
          'verbosity': -1,
          'objective': 'gamma'}

# Train 3 models with the same parameters except nthreads
params["nthreads"] = 8
model_8 = lightgbm.train(params,
                       train_data_lgb,
                       valid_sets = [train_data_lgb, valid_data_lgb],
                       valid_names = ['train', 'valid'],
                       verbose_eval = False)
params["nthreads"] = 4
model_4 = lightgbm.train(params,
                       train_data_lgb,
                       valid_sets = [train_data_lgb, valid_data_lgb],
                       valid_names = ['train', 'valid'],
                       verbose_eval = False)
params["nthreads"] = 1
model_1 = lightgbm.train(params,
                       train_data_lgb,
                       valid_sets = [train_data_lgb, valid_data_lgb],
                       valid_names = ['train', 'valid'],
                       verbose_eval = False)

# Models are differents in terms of performance
[model_8.best_score,model_4.best_score,model_1.best_score]

tree_1 = model_1.trees_to_dataframe()
tree_4 = model_4.trees_to_dataframe()
tree_8 = model_8.trees_to_dataframe()

# Get numbers of sample selected by bagging at each iteration between iteration 10-20,
# which is the sample count at each root node because 1 tree is built per iteration.
# They are also different across models.
tree_1[tree_1["node_depth"]==1]['count'][10:20]
tree_4[tree_4["node_depth"]==1]['count'][10:20]
tree_8[tree_8["node_depth"]==1]['count'][10:20]
`

Output:
`[defaultdict(collections.OrderedDict,
             {'train': OrderedDict([('gamma', 0.9135329616857876)]),
              'valid': OrderedDict([('gamma', 0.9135329616857876)])}),
 defaultdict(collections.OrderedDict,
             {'train': OrderedDict([('gamma', 0.9137869888188606)]),
              'valid': OrderedDict([('gamma', 0.9137869888188606)])}),
 defaultdict(collections.OrderedDict,
             {'train': OrderedDict([('gamma', 0.9137198618188539)]),
              'valid': OrderedDict([('gamma', 0.9137198618188539)])})]

148    3647
167    3647
182    3647
203    3647
220    3647
237    3647
254    3647
271    3647
288    3647
307    3647
Name: count, dtype: int64

148    3644
167    3644
180    3644
197    3644
214    3644
231    3644
242    3644
263    3644
280    3644
289    3644
Name: count, dtype: int64

148    3640
165    3640
180    3640
203    3640
218    3640
237    3640
252    3640
271    3640
292    3640
315    3640
Name: count, dtype: int64`
